Zotero/Controllers/Sync/SyncActions/UploadFixSyncAction.swift

```diff
@@ -0,0 +1,168 @@
+            }
+
+            return Disposables.create()
+        }
+    }
+
+    private func download(attachment: Attachment) -> Single<()> {
+        return Single.create { [weak self] subscriber in
+            guard let self else {
+                subscriber(.failure(Error.expired))
+                return Disposables.create()
+            }
+
+            self.finishDownload = subscriber
+            self.attachmentDownloader.downloadIfNeeded(attachment: attachment, parentKey: nil)
+            return Disposables.create()
+        }
+    }
+
+    private func fetchAndValidateAttachment() -> Single<Attachment> {
+        return Single.create { [weak self] subscriber in
+            guard let self else {
+                subscriber(.failure(Error.expired))
+                return Disposables.create()
+            }
+
+            do {
+                let item = try self.dbStorage.perform(request: ReadItemDbRequest(libraryId: self.libraryId, key: self.key), on: self.queue)
+
+                guard item.rawType == ItemTypes.attachment, let attachment = AttachmentCreator.attachment(for: item, options: .light, fileStorage: self.fileStorage, urlDetector: nil) else {
+                    DDLogError("UploadFixSyncAction: item not attachment - \(item.rawType)")
+                    subscriber(.failure(Error.itemNotAttachment))
+                    return Disposables.create()
+                }
+
+                switch attachment.type {
+                case .url:
+                    DDLogError("UploadFixSyncAction: incorrect item type - \(attachment.type)")
+                    subscriber(.failure(Error.incorrectAttachmentType(attachment.type)))
+
+                case .file(let filename, let contentType, let location, let linkType):
+                    switch linkType {
+                    case .embeddedImage, .linkedFile:
+                        DDLogError("UploadFixSyncAction: incorrect link type - \(linkType)")
+                        subscriber(.failure(Error.incorrectLinkType(linkType)))
+
+                    case .importedFile, .importedUrl:
+                        switch location {
+                        case .remoteMissing:
+                            DDLogError("UploadFixSyncAction: attachment missing remotely")
+                            subscriber(.failure(Error.attachmentMissingRemotely))
+                            return Disposables.create()
+
+                        case .local, .localAndChangedRemotely, .remote: break
+                        }
+
+                        // Create new attachment model with updated location so that `AttachmentDownloader` doesn't ignore it
+                        let newAttachment = Attachment(type: .file(filename: filename, contentType: contentType, location: .remote, linkType: linkType),
+                                                       title: attachment.title, key: attachment.key, libraryId: attachment.libraryId)
+                        subscriber(.success(newAttachment))
```

Comment:
This seems a bit more clear, since switch is exhaustive.

```suggestion

                        case .local, .localAndChangedRemotely, .remote:
                            // Create new attachment model with updated location so that `AttachmentDownloader` doesn't ignore it
                            let newAttachment = Attachment(type: .file(filename: filename, contentType: contentType, location: .remote, linkType: linkType),
                                                           title: attachment.title, key: attachment.key, libraryId: attachment.libraryId)
                            subscriber(.success(newAttachment))
                        }
```

Reply:
Check abfcf1ab5443c69ed6e054242a368773a2bdecc0